File: python/statusrules/engine.py
"""
engine.py — shipment status transition rule engine.

Algorithm overview
------------------
The adjacency table is a flat array of unsigned-short (uint16) flags with
8 192 entries, matching the Go implementation exactly.

Key packing
~~~~~~~~~~~
    key = (from_id << 7) | (to_id << 1) | flow_bit

where:
  * from_id / to_id are Python IDs (yaml_id + 1), each fitting in 6 bits
    (values 1..55 ≤ 63).
  * flow_bit: FORWARD → 0, REVERSE → 1
  * The table therefore needs  64 × 64 × 2 = 8 192 slots.

Edge flags (uint16 bitmask)
~~~~~~~~~~~~~~~~~~~~~~~~~~~
  0x01  FLAG_PRESENT        — edge exists
  0x02  FLAG_OVERRIDE       — requires operator_id (FORCE_*)
  0x04  FLAG_REQUIRES_OP_ID — alias for FLAG_OVERRIDE (same bit)
  0x08  FLAG_INTER_SYSTEM   — (not set at compile time; derived at runtime)

Thread safety
~~~~~~~~~~~~~
The plain int counters (``_cnt_*``) are updated without locks.  CPython's GIL
makes individual integer increments atomic, so reads via ``snapshot()`` are
safe but may race on rapid-fire concurrent calls.  If you need hard guarantees
under free-threaded Python (PEP 703 / Python 3.13t), replace the counters with
``threading.Lock``-guarded fields.
"""

from __future__ import annotations

import array
from pathlib import Path
from typing import List, Optional

import yaml

from .status import CODE_TO_ID, ID_TO_CODE, MAX_ID
from .types import (
    ErrorCode,
    Flow,
    MetricsSnapshot,
    TransitionContext,
    ValidationResult,
)

# ---------------------------------------------------------------------------
# Edge-flag constants
# ---------------------------------------------------------------------------

FLAG_PRESENT        = 0x01
FLAG_OVERRIDE       = 0x02  # edge carries is_override; operator_id required
FLAG_REQUIRES_OP_ID = 0x04  # same semantic — set alongside FLAG_OVERRIDE
FLAG_INTER_SYSTEM   = 0x08  # derived at validate() time, not stored in table

# ---------------------------------------------------------------------------
# Default spec path (relative to this file so the package is relocatable)
# ---------------------------------------------------------------------------

_DEFAULT_SPEC: Path = (
    Path(__file__).parent.parent.parent / "spec" / "rules.yaml"
)


class Engine:
    """Compiled rule engine for validating shipment status transitions.

    Parameters
    ----------
    spec_path:
        Override the location of ``rules.yaml``.  Defaults to the canonical
        path three directory levels above this file (repo root/spec/rules.yaml).

    Example
    -------
    ::

        engine = Engine()
        result = engine.validate(
            current=Status.HUB_IN_SCANNED,
            next_status=Status.HUB_OUTSCAN,
            flow=Flow.FORWARD,
            ctx=TransitionContext(source_system=SourceSystem.HUB),
        )
        assert result.valid
    """

    # ------------------------------------------------------------------
    # Construction / compilation
    # ------------------------------------------------------------------

    def __init__(self, spec_path: Optional[Path] = None) -> None:
        path = Path(spec_path) if spec_path is not None else _DEFAULT_SPEC
        raw = path.read_text(encoding="utf-8")
        spec = yaml.safe_load(raw)
        self._compile(spec)

    def _compile(self, spec: dict) -> None:
        """Build the adjacency array and supporting metadata from the parsed
        YAML spec dict.  Called exactly once from ``__init__``."""

        # ── 1. Build code → python-id mapping ────────────────────────────────
        # yaml_id is 0-indexed; python id = yaml_id + 1.
        code_to_id: dict[str, int] = {}
        for entry in spec["statuses"]:
            code_to_id[entry["code"]] = entry["id"] + 1

        # Sanity: must agree with the static CODE_TO_ID table.
        assert code_to_id == CODE_TO_ID, (
            "Loaded spec does not match the static CODE_TO_ID table in "
            "status.py — the spec may have been renumbered."
        )

        # ── 2. owners[id] = owned_by string  (index 0 unused) ────────────────
        owners: List[str] = [''] * (MAX_ID + 1)
        terminals: List[bool] = [False] * (MAX_ID + 1)
        for entry in spec["statuses"]:
            pid = entry["id"] + 1
            owners[pid] = entry.get("owned_by", '')
            terminals[pid] = bool(entry.get("terminal", False))

        # ── 3. Adjacency array (8 192 unsigned-short slots) ───────────────────
        # Initialised to all-zeros; a zero entry means "no edge".
        adj: array.array = array.array('H', bytes(8192 * 2))

        def _set(from_id: int, to_id: int, flow: Flow, flags: int) -> None:
            """OR *flags* into the adjacency slot for (from_id, to_id, flow)."""
            flow_bit = 0 if flow == Flow.FORWARD else 1
            key = (from_id << 7) | (to_id << 1) | flow_bit
            adj[key] = adj[key] | flags

        # ── 4. Expand wildcard targets ────────────────────────────────────────
        # A wildcard target is reachable from every non-terminal status in
        # BOTH flows.
        wildcard_ids = {
            code_to_id[code]
            for code in spec.get("wildcard_targets", [])
        }
        for entry in spec["statuses"]:
            from_id = entry["id"] + 1
            if terminals[from_id]:
                continue  # terminal statuses have no outbound edges
            for to_id in wildcard_ids:
                for flow in (Flow.FORWARD, Flow.REVERSE):
                    _set(from_id, to_id, flow, FLAG_PRESENT)

        # ── 5. Apply explicit transitions ─────────────────────────────────────
        def _resolve(code_or_list) -> List[int]:
            """Normalise a ``to`` value (string or list of strings) to a list
            of Python status IDs."""
            if isinstance(code_or_list, list):
                return [code_to_id[c] for c in code_or_list]
            return [code_to_id[code_or_list]]

        for txn in spec.get("transitions", []):
            from_id = code_to_id[txn["from"]]
            flow    = Flow.FORWARD if txn["flow"] == "FORWARD" else Flow.REVERSE
            flags   = FLAG_PRESENT

            if txn.get("is_override"):
                flags |= FLAG_OVERRIDE | FLAG_REQUIRES_OP_ID

            for to_id in _resolve(txn["to"]):
                _set(from_id, to_id, flow, flags)

        # ── 6. Build initial-statuses set ────────────────────────────────────
        initials: set[tuple[int, Flow]] = set()
        for entry in spec.get("initial_statuses", []):
            sid   = code_to_id[entry["code"]]
            fflow = Flow.FORWARD if entry["flow"] == "FORWARD" else Flow.REVERSE
            initials.add((sid, fflow))

        # ── Store compiled data ───────────────────────────────────────────────
        self._adj      = adj
        self._owners   = owners
        self._terminals = terminals
        self._initials = initials
        self._version  = spec.get("spec_version", "")

        # Metrics counters (GIL-atomic for CPython; see module docstring).
        self._cnt_total       = 0
        self._cnt_invalid     = 0
        self._cnt_override    = 0
        self._cnt_inter       = 0
        self._cnt_zero        = 0

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def validate(
        self,
        current: int,
        next_status: int,
        flow: Flow,
        ctx: TransitionContext,
    ) -> ValidationResult:
        """Validate a status transition for an in-flight shipment.

        Parameters
        ----------
        current:     Python status ID of the *current* status.
        next_status: Python status ID of the *proposed next* status.
        flow:        ``Flow.FORWARD`` or ``Flow.REVERSE``.
        ctx:         Caller context (source system, operator_id, AWB).

        Returns
        -------
        ValidationResult with ``valid=True`` when the transition is allowed.
        """
        self._cnt_total += 1

        # ── Zero-value guard ──────────────────────────────────────────────────
        if current == 0 or next_status == 0:
            self._cnt_zero    += 1
            self._cnt_invalid += 1
            return ValidationResult(
                valid=False,
                error_code=ErrorCode.ZERO_VALUE_STATUS,
                anomaly_reason="current or next_status is the zero sentinel",
            )

        # ── Unknown-status guard ──────────────────────────────────────────────
        if current > MAX_ID or next_status > MAX_ID:
            self._cnt_invalid += 1
            return ValidationResult(
                valid=False,
                error_code=ErrorCode.UNKNOWN_STATUS,
                anomaly_reason=(
                    f"status id out of range: current={current} "
                    f"next={next_status} max={MAX_ID}"
                ),
            )

        # ── Terminal guard ────────────────────────────────────────────────────
        if self._terminals[current]:
            self._cnt_invalid += 1
            return ValidationResult(
                valid=False,
                error_code=ErrorCode.TERMINAL_STATUS,
                anomaly_reason=(
                    f"{ID_TO_CODE.get(current, current)} is a terminal status"
                ),
            )

        # ── Adjacency lookup ──────────────────────────────────────────────────
        flow_bit = 0 if flow == Flow.FORWARD else 1
        key   = (current << 7) | (next_status << 1) | flow_bit
        flags = self._adj[key]

        if not (flags & FLAG_PRESENT):
            self._cnt_invalid += 1
            return ValidationResult(
                valid=False,
                error_code=ErrorCode.INVALID_TRANSITION,
                anomaly_reason=(
                    f"no edge {ID_TO_CODE.get(current, current)} → "
                    f"{ID_TO_CODE.get(next_status, next_status)} "
                    f"in flow {flow.name}"
                ),
            )

        # ── Override / operator_id check ─────────────────────────────────────
        is_override = bool(flags & FLAG_OVERRIDE)
        if is_override and not ctx.operator_id:
            self._cnt_invalid += 1
            return ValidationResult(
                valid=False,
                is_override=True,
                error_code=ErrorCode.MISSING_OPERATOR_ID,
                anomaly_reason=(
                    f"transition to {ID_TO_CODE.get(next_status, next_status)} "
                    "requires operator_id in context"
                ),
            )

        # ── Inter-system detection ────────────────────────────────────────────
        owner = self._owners[next_status]
        is_inter = bool(
            ctx.source_system
            and owner
            and ctx.source_system != owner
        )

        # ── Metrics ───────────────────────────────────────────────────────────
        if is_override:
            self._cnt_override += 1
        if is_inter:
            self._cnt_inter += 1

        return ValidationResult(
            valid=True,
            is_override=is_override,
            is_inter_system=is_inter,
            error_code=ErrorCode.NONE,
        )

    def validate_initial(
        self,
        first: int,
        flow: Flow,
        ctx: TransitionContext,
    ) -> ValidationResult:
        """Validate the *first* status for a brand-new shipment.

        Unlike ``validate``, there is no *current* status — only a proposed
        *first* status.  The engine checks the ``initial_statuses`` list from
        the spec rather than the adjacency table.

        Parameters
        ----------
        first: Python status ID of the proposed initial status.
        flow:  ``Flow.FORWARD`` or ``Flow.REVERSE``.
        ctx:   Caller context (used for inter-system and zero-value checks).
        """
        self._cnt_total += 1

        if first == 0:
            self._cnt_zero    += 1
            self._cnt_invalid += 1
            return ValidationResult(
                valid=False,
                error_code=ErrorCode.ZERO_VALUE_STATUS,
                anomaly_reason="first status is the zero sentinel",
            )

        if first > MAX_ID:
            self._cnt_invalid += 1
            return ValidationResult(
                valid=False,
                error_code=ErrorCode.UNKNOWN_STATUS,
                anomaly_reason=f"status id out of range: first={first} max={MAX_ID}",
            )

        if (first, flow) not in self._initials:
            self._cnt_invalid += 1
            return ValidationResult(
                valid=False,
                error_code=ErrorCode.INVALID_TRANSITION,
                anomaly_reason=(
                    f"{ID_TO_CODE.get(first, first)} / {flow.name} "
                    "is not a valid initial status"
                ),
            )

        owner    = self._owners[first]
        is_inter = bool(
            ctx.source_system
            and owner
            and ctx.source_system != owner
        )
        if is_inter:
            self._cnt_inter += 1

        return ValidationResult(
            valid=True,
            is_inter_system=is_inter,
            error_code=ErrorCode.NONE,
        )

    def permitted_transitions(self, current: int, flow: Flow) -> List[int]:
        """Return a sorted list of Python status IDs reachable from *current*
        in the given *flow*.

        Parameters
        ----------
        current: Python status ID of the source status.
        flow:    ``Flow.FORWARD`` or ``Flow.REVERSE``.

        Returns
        -------
        List of integer status IDs (may be empty if *current* is terminal or
        has no outbound edges for this flow).
        """
        if current <= 0 or current > MAX_ID:
            return []
        if self._terminals[current]:
            return []

        flow_bit = 0 if flow == Flow.FORWARD else 1
        result: List[int] = []
        for to_id in range(1, MAX_ID + 1):
            key = (current << 7) | (to_id << 1) | flow_bit
            if self._adj[key] & FLAG_PRESENT:
                result.append(to_id)
        return result

    def spec_version(self) -> str:
        """Return the ``spec_version`` string from the loaded YAML (e.g. ``'1.0.0'``)."""
        return self._version

    def snapshot(self) -> MetricsSnapshot:
        """Return a point-in-time copy of the engine's cumulative counters.

        Thread safety: each attribute read is GIL-atomic in CPython, but two
        successive reads are not.  Under concurrent load the snapshot may be
        slightly inconsistent across fields.  This is acceptable for
        monitoring/telemetry use cases.
        """
        return MetricsSnapshot(
            validations_total          = self._cnt_total,
            invalid_transitions        = self._cnt_invalid,
            override_transitions       = self._cnt_override,
            inter_system_transitions   = self._cnt_inter,
            zero_value_rejections      = self._cnt_zero,
        )

File: java/src/main/java/io/logistics/statusrules/Engine.java
package io.logistics.statusrules;

import org.yaml.snakeyaml.Yaml;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Core rule engine for shipment status transitions.
 *
 * <h2>Adjacency table layout</h2>
 * <p>The engine compiles the YAML spec into a flat {@code short[8192]} array.
 * Each entry is addressed by a packed integer key:
 * <pre>
 *   key = (fromId &lt;&lt; 7) | (toId &lt;&lt; 1) | flowBit
 *   flowBit: FORWARD = 0, REVERSE = 1
 * </pre>
 * Status IDs are the YAML 0-based {@code id} field plus one (matching the Go
 * convention where 0 is the uninitialised sentinel). With MAX_ID = 55, the
 * maximum key is (55 &lt;&lt; 7) | (55 &lt;&lt; 1) | 1 = 7040 + 110 + 1 = 7151, which
 * fits comfortably in the 8192-entry array.
 *
 * <h2>Edge flags</h2>
 * <pre>
 *   FLAG_PRESENT       = 0x01  — edge exists
 *   FLAG_OVERRIDE      = 0x02  — FORCE_* / override transition; requires operatorId
 *   FLAG_REQUIRES_OPID = 0x04  — alias for override requirement (same bit field)
 *   FLAG_INTER_SYSTEM  = 0x08  — from-owner ≠ to-owner
 * </pre>
 *
 * <h2>Thread safety</h2>
 * <p>After construction the compiled tables are immutable. Metric counters use
 * {@link AtomicLong} so the engine is safe to share across threads.
 */
public final class Engine {

    // ── Adjacency-table constants ──────────────────────────────────────────────

    /** Total adjacency-table entries: IDs 0-63 × IDs 0-63 × 2 flows = 8192. */
    private static final int ADJ_SIZE = 8192;

    /** Bit 0 — the edge exists in the spec. */
    static final short FLAG_PRESENT       = 0x01;

    /** Bit 1 — this is an override (FORCE_*) edge. */
    static final short FLAG_OVERRIDE      = 0x02;

    /** Bit 2 — the edge requires a non-blank operatorId in the context. */
    static final short FLAG_REQUIRES_OPID = 0x04;

    /** Bit 3 — the source system does not own the destination status. */
    static final short FLAG_INTER_SYSTEM  = 0x08;

    // ── Compiled spec tables ───────────────────────────────────────────────────

    /**
     * Packed adjacency table. Each element is a bitset of FLAG_* constants.
     * Java {@code short} is 16-bit signed; we read it as unsigned via
     * {@code & 0xFFFF} wherever comparison is needed.
     */
    private final short[] adj;

    /**
     * Per-status owning system (indexed by status ID, 1-based).
     * Element [0] is unused (zero-value sentinel).
     */
    private final SourceSystem[] owners;

    /**
     * Per-status terminal flag (indexed by status ID, 1-based).
     * Element [0] is unused.
     */
    private final boolean[] terminals;

    /**
     * Per-status flag: true if this status has category FIRST_MILE.
     * Element [0] is unused.
     */
    private final boolean[] firstMile;

    /**
     * Per-status flag: true if this status has category MIDDLE_MILE or LAST_MILE.
     * Element [0] is unused.
     */
    private final boolean[] postFirstMile;

    /**
     * Valid first statuses for new shipments.
     * Each element is a two-element array {@code [statusId, flowBit]}.
     */
    private final List<int[]> initials;

    /** Spec version string read from the YAML {@code spec_version} field. */
    private final String specVersion;

    // ── Metrics counters ───────────────────────────────────────────────────────

    private final AtomicLong validationsTotal      = new AtomicLong();
    private final AtomicLong invalidTransitions     = new AtomicLong();
    private final AtomicLong overrideTransitions    = new AtomicLong();
    private final AtomicLong interSystemTransitions = new AtomicLong();
    private final AtomicLong zeroValueRejections    = new AtomicLong();

    // ── Constructor ────────────────────────────────────────────────────────────

    /** All construction goes through the static factories. */
    private Engine(short[] adj,
                   SourceSystem[] owners,
                   boolean[] terminals,
                   boolean[] firstMile,
                   boolean[] postFirstMile,
                   List<int[]> initials,
                   String specVersion) {
        this.adj           = adj;
        this.owners        = owners;
        this.terminals     = terminals;
        this.firstMile     = firstMile;
        this.postFirstMile = postFirstMile;
        this.initials      = Collections.unmodifiableList(initials);
        this.specVersion   = specVersion;
    }

    // ── Static factories ───────────────────────────────────────────────────────

    /**
     * Creates an engine by reading {@code ../spec/rules.yaml} relative to the
     * {@code java/} working directory.  This is the default production factory.
     *
     * @return a compiled, ready-to-use engine
     * @throws IOException if the spec file cannot be read
     */
    public static Engine newEngine() throws IOException {
        return newEngineFromPath(Path.of("../spec/rules.yaml"));
    }

    /**
     * Creates an engine from an explicit file path.  Useful for testing or
     * deployments where the spec lives at a non-standard location.
     *
     * @param specPath absolute or relative path to the YAML spec file
     * @return a compiled engine
     * @throws IOException if the file cannot be read
     */
    public static Engine newEngineFromPath(Path specPath) throws IOException {
        byte[] bytes = Files.readAllBytes(specPath);
        return newEngineFromSpec(bytes);
    }

    /**
     * Creates an engine from raw YAML bytes.  Useful for tests that supply the
     * spec as an in-memory resource rather than a file on disk.
     *
     * @param yaml raw UTF-8 bytes of the rules YAML
     * @return a compiled engine
     */
    public static Engine newEngineFromSpec(byte[] yaml) {
        return compileSpec(yaml);
    }

    // ── Validation API ─────────────────────────────────────────────────────────

    /**
     * Validates whether transitioning from {@code current} to {@code next}
     * under the given {@code flow} is permitted for the supplied context.
     *
     * <p>Call order:
     * <ol>
     *   <li>Null / zero-value guard
     *   <li>Terminal check on {@code current}
     *   <li>Adjacency lookup
     *   <li>Override / operatorId check
     *   <li>Inter-system detection
     * </ol>
     *
     * @param current the shipment's present status
     * @param next    the desired next status
     * @param flow    direction of the lifecycle
     * @param ctx     caller context (source system, operator)
     * @return a {@link ValidationResult} describing the outcome
     */
    public ValidationResult validate(Status current,
                                     Status next,
                                     Flow flow,
                                     TransitionContext ctx) {
        validationsTotal.incrementAndGet();

        // 1. Zero-value / null guard
        if (current == null || next == null) {
            zeroValueRejections.incrementAndGet();
            invalidTransitions.incrementAndGet();
            return ValidationResult.error(
                    ErrorCode.ZERO_VALUE_STATUS,
                    "current and next status must both be non-null");
        }

        // 2. Terminal check — no transitions out of a terminal status
        if (terminals[current.id]) {
            invalidTransitions.incrementAndGet();
            return ValidationResult.error(
                    ErrorCode.TERMINAL_STATUS,
                    current.code + " is a terminal status; no further transitions are allowed");
        }

        // 2b. First-mile regression check
        if (postFirstMile[current.id] && firstMile[next.id]) {
            invalidTransitions.incrementAndGet();
            return ValidationResult.error(
                    ErrorCode.FIRST_MILE_AFTER_LATER_MILE,
                    "first-mile status " + next.code + " cannot follow middle-mile or last-mile status " + current.code);
        }

        // 3. Adjacency lookup
        int key       = packKey(current.id, next.id, flow);
        int flags     = adj[key] & 0xFFFF;

        if ((flags & FLAG_PRESENT) == 0) {
            invalidTransitions.incrementAndGet();
            return ValidationResult.error(
                    ErrorCode.INVALID_TRANSITION,
                    current.code + " → " + next.code + " (" + flow + ") is not a permitted transition");
        }

        // 4. Override / operatorId requirement
        boolean requiresOpId = (flags & FLAG_REQUIRES_OPID) != 0;
        boolean isOverride   = (flags & FLAG_OVERRIDE)      != 0;

        if (requiresOpId && (ctx.operatorId() == null || ctx.operatorId().isBlank())) {
            invalidTransitions.incrementAndGet();
            return ValidationResult.error(
                    ErrorCode.MISSING_OPERATOR_ID,
                    "Transition " + current.code + " → " + next.code
                            + " requires a non-blank operatorId in the context");
        }

        // 5. Inter-system detection
        boolean isInterSystem = (flags & FLAG_INTER_SYSTEM) != 0;

        // Update metrics
        if (isOverride)     overrideTransitions.incrementAndGet();
        if (isInterSystem)  interSystemTransitions.incrementAndGet();

        // Build result
        if (isOverride && isInterSystem) return ValidationResult.okOverrideInterSystem();
        if (isOverride)                  return ValidationResult.okOverride();
        if (isInterSystem)               return ValidationResult.okInterSystem();
        return ValidationResult.ok();
    }

    /**
     * Validates that {@code first} is a permissible initial status for a new
     * shipment (i.e. no previous status exists).
     *
     * @param first the proposed first status
     * @param flow  direction of the lifecycle
     * @param ctx   caller context
     * @return a {@link ValidationResult} describing the outcome
     */
    public ValidationResult validateInitial(Status first, Flow flow, TransitionContext ctx) {
        validationsTotal.incrementAndGet();

        if (first == null) {
            zeroValueRejections.incrementAndGet();
            invalidTransitions.incrementAndGet();
            return ValidationResult.error(
                    ErrorCode.ZERO_VALUE_STATUS,
                    "first status must be non-null");
        }

        int flowBit = (flow == Flow.REVERSE) ? 1 : 0;

        for (int[] entry : initials) {
            if (entry[0] == first.id && entry[1] == flowBit) {
                return ValidationResult.ok();
            }
        }

        invalidTransitions.incrementAndGet();
        return ValidationResult.error(
                ErrorCode.INVALID_TRANSITION,
                first.code + " (" + flow + ") is not a valid initial status");
    }

    /**
     * Returns all statuses that are reachable from {@code current} under
     * {@code flow} according to the compiled spec.
     *
     * @param current the shipment's present status
     * @param flow    direction of the lifecycle
     * @return an immutable list of permitted next statuses (may be empty)
     */
    public List<Status> permittedTransitions(Status current, Flow flow) {
        if (current == null || terminals[current.id]) {
            return Collections.emptyList();
        }

        List<Status> result = new ArrayList<>();
        int flowBit = (flow == Flow.REVERSE) ? 1 : 0;

        for (Status next : Status.values()) {
            int key   = packKey(current.id, next.id, flow);
            int flags = adj[key] & 0xFFFF;
            if ((flags & FLAG_PRESENT) != 0) {
                result.add(next);
            }
        }

        return Collections.unmodifiableList(result);
    }

    // ── Metadata & metrics ────────────────────────────────────────────────────

    /**
     * Returns the {@code spec_version} string from the compiled YAML spec.
     *
     * @return spec version, e.g. {@code "1.0.0"}
     */
    public String specVersion() {
        return specVersion;
    }

    /**
     * Returns a point-in-time snapshot of all validation counters.
     *
     * @return an immutable {@link MetricsSnapshot}
     */
    public MetricsSnapshot snapshot() {
        return new MetricsSnapshot(
                validationsTotal.get(),
                invalidTransitions.get(),
                overrideTransitions.get(),
                interSystemTransitions.get(),
                zeroValueRejections.get());
    }

    // ── Spec compilation ───────────────────────────────────────────────────────

    /**
     * Compiles the YAML spec into the in-memory adjacency table.
     *
     * <p><b>Algorithm (mirrors Go implementation):</b>
     * <ol>
     *   <li>Parse YAML into a raw {@code Map<String, Object>}.
     *   <li>Build a {@code code → Status} lookup and populate {@code owners[]}
     *       and {@code terminals[]}.
     *   <li>Collect wildcard-target codes and initial-status entries.
     *   <li>For each transition entry, resolve the {@code from} code, normalise
     *       the {@code to} field (String or List<String>), and write edge flags.
     *   <li>For every non-terminal status and every wildcard target, write a
     *       wildcard edge in both flows.
     * </ol>
     */
    @SuppressWarnings("unchecked")
    private static Engine compileSpec(byte[] yaml) {

        // ── 1. Parse YAML ──────────────────────────────────────────────────────
        Yaml parser = new Yaml();
        Map<String, Object> root = parser.load(new String(yaml));

        String specVersion = (String) root.get("spec_version");
        if (specVersion == null) specVersion = "";

        // ── 2. Build status registry ───────────────────────────────────────────
        // code→id map (YAML ids are 0-based; we store them as id+1)
        Map<String, Integer> codeToId = new HashMap<>();
        // Indexed by Java id (1-based)
        SourceSystem[] owners       = new SourceSystem[Status.MAX_ID + 1];
        boolean[]      terminals    = new boolean[Status.MAX_ID + 1];
        boolean[]      firstMile    = new boolean[Status.MAX_ID + 1];
        boolean[]      postFirstMile = new boolean[Status.MAX_ID + 1];

        List<Map<String, Object>> statuses =
                (List<Map<String, Object>>) root.get("statuses");

        for (Map<String, Object> entry : statuses) {
            String code   = (String) entry.get("code");
            int    yamlId = (Integer) entry.get("id");
            int    javaId = yamlId + 1;           // +1: 0 is the zero-value sentinel

            codeToId.put(code, javaId);

            String ownedBy = (String) entry.get("owned_by");
            owners[javaId] = SourceSystem.fromYaml(ownedBy);

            Object terminalObj = entry.get("terminal");
            terminals[javaId]  = Boolean.TRUE.equals(terminalObj);

            String category = entry.containsKey("category") ? ((String) entry.get("category")).toUpperCase() : "";
            if ("FIRST_MILE".equals(category)) {
                firstMile[javaId] = true;
            } else if ("MIDDLE_MILE".equals(category) || "LAST_MILE".equals(category)) {
                postFirstMile[javaId] = true;
            }
        }

        // ── 3. Wildcard targets ────────────────────────────────────────────────
        List<String> wildcardCodes = new ArrayList<>();
        Object wildcardRaw = root.get("wildcard_targets");
        if (wildcardRaw instanceof List<?> wcList) {
            for (Object wc : wcList) {
                wildcardCodes.add((String) wc);
            }
        }

        // ── 4. Initial statuses ────────────────────────────────────────────────
        List<int[]> initials = new ArrayList<>();
        Object initialRaw = root.get("initial_statuses");
        if (initialRaw instanceof List<?> initList) {
            for (Object item : initList) {
                Map<String, Object> initEntry = (Map<String, Object>) item;
                String code = (String) initEntry.get("code");
                String flowStr = (String) initEntry.get("flow");
                Integer id = codeToId.get(code);
                if (id == null) continue;
                int flowBit = "REVERSE".equals(flowStr) ? 1 : 0;
                initials.add(new int[]{id, flowBit});
            }
        }

        // ── 5. Build adjacency table ───────────────────────────────────────────
        short[] adj = new short[ADJ_SIZE];

        List<Map<String, Object>> transitions =
                (List<Map<String, Object>>) root.get("transitions");

        for (Map<String, Object> tx : transitions) {
            String fromCode = (String) tx.get("from");
            Integer fromId  = codeToId.get(fromCode);
            if (fromId == null) continue;                // defensive: unknown status

            // 'flow' field maps to a Flow enum
            String flowStr = (String) tx.get("flow");
            Flow flow      = "REVERSE".equals(flowStr) ? Flow.REVERSE : Flow.FORWARD;

            // Flags on this edge
            boolean isOverride   = Boolean.TRUE.equals(tx.get("is_override"));
            boolean requiresOpId = false;

            Object ctxRaw = tx.get("requires_context");
            if (ctxRaw instanceof List<?> ctxList) {
                for (Object c : ctxList) {
                    if ("operator_id".equals(c)) {
                        requiresOpId = true;
                    }
                }
            }

            // 'to' may be a single String or a List<String>
            List<String> toCodes = new ArrayList<>();
            Object toRaw = tx.get("to");
            if (toRaw instanceof String singleTo) {
                toCodes.add(singleTo);
            } else if (toRaw instanceof List<?> toList) {
                for (Object t : toList) {
                    toCodes.add((String) t);
                }
            }

            for (String toCode : toCodes) {
                Integer toId = codeToId.get(toCode);
                if (toId == null) continue;              // defensive: unknown status

                // Determine inter-system flag
                boolean interSystem = !owners[fromId].equals(owners[toId]);

                // Compute flags
                short flags = FLAG_PRESENT;
                if (isOverride)   flags |= FLAG_OVERRIDE;
                if (requiresOpId) flags |= FLAG_REQUIRES_OPID;
                if (interSystem)  flags |= FLAG_INTER_SYSTEM;

                int key = packKey(fromId, toId, flow);
                adj[key] |= flags;
            }
        }

        // ── 6. Expand wildcards ────────────────────────────────────────────────
        // Every non-terminal status may transition to any wildcard target in
        // either flow direction.
        for (int fromId = 1; fromId <= Status.MAX_ID; fromId++) {
            if (terminals[fromId]) continue;

            for (String wcCode : wildcardCodes) {
                Integer toId = codeToId.get(wcCode);
                if (toId == null) continue;

                boolean interSystem = !owners[fromId].equals(owners[toId]);
                short   flags       = FLAG_PRESENT;
                if (interSystem) flags |= FLAG_INTER_SYSTEM;

                // Apply in both directions
                for (Flow f : Flow.values()) {
                    int key = packKey(fromId, toId, f);
                    adj[key] |= flags;
                }
            }
        }

        return new Engine(adj, owners, terminals, firstMile, postFirstMile, initials, specVersion);
    }

    // ── Key packing ────────────────────────────────────────────────────────────

    /**
     * Packs a (from, to, flow) triple into a single integer adjacency-table key.
     *
     * <pre>
     *   key = (fromId &lt;&lt; 7) | (toId &lt;&lt; 1) | flowBit
     * </pre>
     *
     * With IDs in the range [1, 55]:
     * <ul>
     *   <li>{@code fromId << 7} uses bits 7–12 (max = 55 × 128 = 7040)
     *   <li>{@code toId   << 1} uses bits 1–6  (max = 55 × 2   =  110)
     *   <li>{@code flowBit}     uses bit  0    (0 = FORWARD, 1 = REVERSE)
     * </ul>
     * Maximum key = 7040 + 110 + 1 = 7151 &lt; 8192 — always in bounds.
     *
     * @param fromId  source status ID (1-based)
     * @param toId    destination status ID (1-based)
     * @param flow    transition direction
     * @return packed key suitable as an index into {@link #adj}
     */
    private static int packKey(int fromId, int toId, Flow flow) {
        int flowBit = (flow == Flow.REVERSE) ? 1 : 0;
        return (fromId << 7) | (toId << 1) | flowBit;
    }
}
